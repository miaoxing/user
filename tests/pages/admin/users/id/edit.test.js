import Form from '../../../../../pages/admin/users/[id]/edit';
import {fireEvent, render, screen} from '@testing-library/react';
import {MemoryRouter} from 'react-router';
import React from 'react';
import {app} from '@mxjs/app';
import $ from 'miaoxing';
import {bootstrap, createPromise, setUrl, resetUrl} from '@mxjs/test';

bootstrap();

describe('admin/users', () => {
  beforeEach(() => {
    setUrl('admin/users/1/edit');
    app.page = {
      collection: 'admin/users',
      index: false,
    };
  });

  afterEach(() => {
    resetUrl();
    app.page = {};
  });

  test('form', async () => {
    const promise = createPromise();
    const promise2 = createPromise();

    $.http = jest.fn()
      // 读取默认数据
      .mockImplementationOnce(() => promise.resolve({
        code: 1,
        data: {
          id: 1,
          name: '姓名2',
          mobile: '13800138000',
          isMobileVerified: true,
        },
      }))
      // 提交
      .mockImplementationOnce(() => promise2.resolve({
        code: 1,
      }));

    const {container, getByLabelText, findByDisplayValue} = render(<MemoryRouter>
      <Form/>
    </MemoryRouter>);

    await Promise.all([promise]);
    expect($.http).toHaveBeenCalledTimes(1);
    expect($.http).toMatchSnapshot();

    // 看到表单加载了数据
    await findByDisplayValue('姓名2');
    await findByDisplayValue('13800138000');
    expect(container.querySelector('#isMobileVerified').checked).toBeTruthy();

    // 提交表单
    fireEvent.change(getByLabelText('姓名'), {target: {value: '新名称'}});
    fireEvent.click(screen.getByText('提 交'));

    await Promise.all([promise2]);
    expect($.http).toHaveBeenCalledTimes(2);
    expect($.http).toMatchSnapshot();
  });
});
