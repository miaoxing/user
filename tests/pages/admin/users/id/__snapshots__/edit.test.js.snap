// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`admin/users form 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "method": "get",
        "url": "/admin-api/users/defaults",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
  ],
}
`;

exports[`admin/users form 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "method": "get",
        "url": "/admin-api/users/defaults",
      },
    ],
    Array [
      Object {
        "data": Object {
          "id": 1,
          "isMobileVerified": true,
          "mobile": "13800138000",
          "name": "新名称",
        },
        "loading": true,
        "method": "POST",
        "url": "/admin-api/users",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
  ],
}
`;
